src/main/java/com/google/api/codegen/metacode/FieldStructureParser.java

```diff
@@ -103,33 +104,92 @@ private static InitCodeNode parsePartialDottedPathToInitCodeNode(
           break;
 
         case '{':
-          token = scanner.scan();
-          Preconditions.checkArgument(
-              token == Scanner.INT || token == Scanner.IDENT || token == Scanner.STRING,
-              "invalid key after '{': %s",
-              path);
+          child = InitCodeNode.create(parseValue(scanner));
           parent.setLineType(InitCodeLineType.MapInitLine);
-          child = InitCodeNode.create(scanner.token());
           parent.mergeChild(child);
           parent = child;
 
-          Preconditions.checkArgument(scanner.scan() == '}', "expected closing '}': %s", path);
+          Preconditions.checkArgument(scanner.scan() == '}', "expected closing '}': %s", config);
           break;
 
         default:
           throw new IllegalArgumentException(
-              String.format("unexpected character '%c': %s", token, path));
+              String.format("unexpected character '%c': %s", token, config));
       }
     }
+
+    int fieldNamePos = config.length();
+
+    String entityName = null;
+    if (token == '%') {
+      fieldNamePos = scanner.pos() - 1;
+      Preconditions.checkArgument(
+          scanner.scan() == Scanner.IDENT, "expected ident after '%': %s", config);
+      entityName = scanner.token();
+      token = scanner.scan();
+    }
+
+    InitValue initValue = null;
+    if (token == '=') {
+      fieldNamePos = Math.min(fieldNamePos, scanner.pos() - 1);
+
+      // TODO(pongad): Currently the RHS of equal sign is just arbitrary run of text treated as string, eg
+      //   a.b=https://foo.bar.com/zip/zap
+      // We'll quote the RHS of existing configs, then we can use parseValue here.
+      // For now, just preserve old behavior.
+      // String valueString = parseValue(scanner);
+      String valueString = config.substring(scanner.pos());
+
+      if (valueString.contains(InitFieldConfig.RANDOM_TOKEN)) {
+        System.err.println(valueString);
+        initValue = InitValue.createRandom(valueString);
+      } else if (valueString.contains(PROJECT_ID_TOKEN)) {
+        Preconditions.checkArgument(
+            valueString.equals(PROJECT_ID_TOKEN),
+            "%s cannot be used as substring: %s",
+            PROJECT_ID_TOKEN,
+            config);
+        initValue = InitValue.createVariable(InitFieldConfig.PROJECT_ID_VARIABLE_NAME);
+      } else {
+        initValue = InitValue.createLiteral(valueString);
+      }
+      token = scanner.scan();
+    }
+
+    // TODO(pongad): When we can actually parse the RHS, we should expect EOF.
+    // Preconditions.checkArgument(scanner.scan() == Scanner.EOF, "expected EOF: %s", config);
+
+    InitValueConfig valueConfig =
+        createInitValueConfig(
+            InitFieldConfig.newBuilder()
+                .fieldPath(config.substring(0, fieldNamePos))
+                .entityName(entityName)
+                .value(initValue)
+                .build(),
+            initValueConfigMap);
+    if (valueConfig != null) {
```

Comment:
What would `valueConfig == null` mean? Should we throw?

Reply:
I don't completely understand this code myself. IIUC, if `valueConfig == null`, somewhere we put in the default value: zero, empty string, empty list, struct with all zero fields, etc.